internal/controller/apiextensions/composite/composition_render.go

```diff
@@ -0,0 +1,194 @@
+	// changed).
+	empty := schema.GroupVersionKind{}
+	if gvk != empty && o.GetObjectKind().GroupVersionKind() != gvk {
+		return errors.Errorf(errFmtKindChanged, gvk, o.GetObjectKind().GroupVersionKind())
+	}
+
+	return nil
+}
+
+// RenderFromCompositePatches renders the supplied composed resource by applying
+// all patches that are _from_ the supplied composite resource.
+func RenderFromCompositePatches(cd resource.Composed, xr resource.Composite, p []v1.Patch) error {
+	for i := range p {
+		if err := Apply(p[i], xr, cd, patchTypesFromXR()...); err != nil {
+			return errors.Wrapf(err, errFmtPatch, p[i].Type, i)
+		}
+	}
+	return nil
+}
+
+// RenderFromEnvironmentPatches renders the supplied composed resource by
+// applying all patches that are from the supplied environment.
+func RenderFromEnvironmentPatches(cd resource.Composed, e *Environment, p []v1.Patch) error {
+	if e == nil {
+		return nil
+	}
+	for i := range p {
+		if err := ApplyToObjects(p[i], e, cd, patchTypesFromToEnvironment()...); err != nil {
+			return errors.Wrapf(err, errFmtPatch, p[i].Type, i)
+		}
+	}
+	return nil
+}
+
+// RenderToCompositePatches renders the supplied composite resource by applying
+// all patches that are _from_ the supplied composed resource. composed resource
+// and template.
+func RenderToCompositePatches(xr resource.Composite, cd resource.Composed, p []v1.Patch) error {
+	for i := range p {
+		if err := Apply(p[i], xr, cd, patchTypesToXR()...); err != nil {
+			return errors.Wrapf(err, errFmtPatch, p[i].Type, i)
+		}
+	}
+	return nil
+}
+
+// RenderComposedResourceMetadata derives composed resource metadata from the
+// supplied composite resource. It makes the composite resource the controller
+// of the composed resource. It should run toward the end of a render pipeline
+// to ensure that a Composition cannot influence the controller reference.
+func RenderComposedResourceMetadata(cd, xr resource.Object, n ResourceName) error {
+	// Fail early if the supplied composite resource is missing the name prefix
+	// label.
+	if xr.GetLabels()[xcrd.LabelKeyNamePrefixForComposed] == "" {
+		return errors.Errorf(errFmtNamePrefixLabel, xcrd.LabelKeyNamePrefixForComposed)
+	}
+
+	//  We also set generate name in case we
+	// haven't yet named this composed resource.
+	cd.SetGenerateName(xr.GetLabels()[xcrd.LabelKeyNamePrefixForComposed] + "-")
+
+	if n != "" {
+		SetCompositionResourceName(cd, n)
+	}
+
+	meta.AddLabels(cd, map[string]string{
+		xcrd.LabelKeyNamePrefixForComposed: xr.GetLabels()[xcrd.LabelKeyNamePrefixForComposed],
+		xcrd.LabelKeyClaimName:             xr.GetLabels()[xcrd.LabelKeyClaimName],
+		xcrd.LabelKeyClaimNamespace:        xr.GetLabels()[xcrd.LabelKeyClaimNamespace],
+	})
+
+	or := meta.AsController(meta.TypedReferenceTo(xr, xr.GetObjectKind().GroupVersionKind()))
+	return errors.Wrap(meta.AddControllerReference(cd, or), errSetControllerRef)
+}
+
+// TODO(negz): Is this really a 'renderer'?
```

Comment:
Yeah, potentially. I do kind of prefer being able to inject interfaces rather than concrete function signatures though, and it's nice to be able to inject this.